system/lib/wasmfs/backends/node_backend.cpp

```diff
@@ -0,0 +1,173 @@
+// Copyright 2022 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+#include <memory>
+
+#include "backend.h"
+#include "file.h"
+#include "wasmfs.h"
+
+namespace wasmfs {
+
+class NodeBackend;
+
+extern "C" {
+
+// Fill `entries` and return 0 or an error code.
+int _wasmfs_node_readdir(const char* path,
+                         std::vector<Directory::Entry>* entries);
+// Write `mode` and return 0 or an error code.
+int _wasmfs_node_get_mode(const char* path, mode_t* mode);
+
+// Create a new file system entry and return 0 or an error code.
+int _wasmfs_node_insert_file(const char* path,
+                             mode_t mode); // fs.openSync(name, flags, mode)
+int _wasmfs_node_insert_directory(
+  const char* path, mode_t mode); // fs.mkdirSync(name, { mode: ... })
+
+} // extern "C"
+
+class NodeFile : public DataFile {
+public:
+  // The path to this file on the underlying Node file system.
+  std::string path;
+
+  NodeFile(mode_t mode, backend_t backend, std::string path)
+    : DataFile(mode, backend), path(path) {
+    // TODO: initialize node file
+  }
+
+private:
+  size_t getSize() override {
+    // TODO: getSize
+    return 0;
```

Comment:
I generally think it's better to put an abort in places like this. Silently returning 0 might happen to be ignored, while an abort is never ignorable. (and an assertion doesn't happen in debug builds, but I could be ok with that too)

Reply:
(can use the new code in #16375 if you want for it to land, or this could be a followup)